### PR TITLE
[unitaryHACK] Update interface introduction

### PR DESCRIPTION
**Pull request by ryanhill1**
**Context:**

Outdated interface introduction

**Description of the Change:**

In the [interface introduction](https://pennylane.readthedocs.io/en/stable/introduction/interfaces/torch.html), updated the sections on converting an existing QNode so that the functions `to_torch()` and `to_tf()` alter the QNode in place instead of returning a new QNode. Also, updated each interface's introduction to show creating a QNode with an interface via the class constructor:

**Benefits:**

**Possible Drawbacks:**

**Related GitHub Issues:**

#1155

### CODE REVIEW

File: doc/introduction/interfaces/tf.rst [hunk 1]
```diff
@@ -5,10 +5,10 @@ TensorFlow interface
 
 In order to use PennyLane in combination with TensorFlow, we have to generate TensorFlow-compatible
 quantum nodes. A basic ``QNode`` can be translated into a quantum node that interfaces with TensorFlow,
-either by using the ``interface='tf'`` flag in the QNode Decorator, or by calling the
-:meth:`QNode.to_tf() <pennylane.qnodes.JacobianQNode.to_tf>` method. Internally, the translation is
-executed by the :func:`~.interfaces.tf.to_tf` function that returns the
-new quantum node object.
+either by using the ``interface='tf'`` flag in the QNode decorator or QNode class constructor, or by
+calling the :meth:`QNode.to_tf() <pennylane.qnodes.JacobianQNode.to_tf>` method. Internally, the
```
Comment (ryanhill1) [inline doc/introduction/interfaces/tf.rst:9]: Is the reference <pennylane.qnodes.JacobianQNode.to_tf> correct? Or should it be <pennylane.qnode.QNode.to_torch> ?

File: doc/introduction/interfaces/torch.rst [hunk 2]
```diff
@@ -68,15 +62,31 @@ use different classical interfaces:
...
+
+    dev2 = qml.device('default.mixed', wires=2)
     qnode2 = qml.QNode(circuit2, dev2)
 
 We can convert the default NumPy-interfacing QNode to a PyTorch-interfacing QNode by
 using the :meth:`~.QNode.to_torch` method:
 
->>> qnode1_torch = qnode1.to_torch()
->>> qnode1_torch
-<QNode: device='default.qubit', func=circuit, wires=2, interface=PyTorch>
```
Comment (ryanhill1) [inline doc/introduction/interfaces/torch.rst:79]: Running the cells as they were before doesn't seem to give the `<QNode: device='default.qubit', func=circuit, wires=2, interface=PyTorch>` output. Instead, I get something like `<pennylane.qnode.QNode at 0x7ff01284d370>`.  Do you know the root of this inconsistency?

File: doc/introduction/interfaces/tf.rst (hunk 1 above)
Comment (albi3ro) [reply doc/introduction/interfaces/tf.rst:9]: The second one.  No such thing as a `JacobianQNode` (I think?).

File: doc/introduction/interfaces/tf.rst [hunk 3]
```diff
@@ -5,10 +5,10 @@ TensorFlow interface
...
+translation is executed by the :func:`~.interfaces.torch.to_torch` function that dynamically modifies
```
Comment (albi3ro) [inline doc/introduction/interfaces/tf.rst:10]: I don't think we need to mention internal details here.

As a note, the interfaces are mix-in classes that can get added onto a tape by `~.interfaces.tf.TFInterface.apply`, but even most developers are not going to be using that information.

**Review by albi3ro — CHANGES_REQUESTED** [review]
Thanks @ryanhill1 !

I left a lot of comments, but I can get pretty opinionated about writing. The comments about flow and phrasing are more optional potential improvements than required changes.

The things that do need to get changed are:
* correct function links and names
* section titles rendering well in build

Potential improvements:
* Remove discussion of implementation details
* explanation of when we may need to convert existing qnode
* move some text around and rephrase some other stuff

**Comment by ryanhill1** [comment]
Thank you so much for your great feedback! I'll get right on those changes

File: doc/introduction/interfaces/tf.rst (hunk 1 above)
Comment (josh146) [reply doc/introduction/interfaces/tf.rst:9]: Yep, `JacobianQNode` is old and has long been removed 🙂

File: doc/introduction/interfaces/torch.rst (hunk 2 above)
Comment (josh146) [reply doc/introduction/interfaces/torch.rst:79]: Good catch, we should definitely add the `__repr__` method back to the QNode - it must have gotten lost in a recent refactor. @albi3ro, could you open an issue for this?

**Comment by albi3ro** [comment]
Looks much better 👍

**Comment by josh146** [comment]
@ryanhill1 is this draft PR now final? 🙂

**Comment by ryanhill1** [comment]
@josh146 Just about! Should I add similar updates to the [JAX interface page](https://github.com/PennyLaneAI/pennylane/blob/master/doc/introduction/interfaces/jax.rst)? I was holding off on that one until I got the PyTorch and TensorFlow ones right so I could just work off of those templates.

File: doc/introduction/interfaces/tf.rst [hunk 4]
```diff
@@ -24,61 +23,78 @@ Tensorflow is imported as follows:
...
+Let's say we change our mind and now want ``qnode1()`` to interface with TensorFlow. We can easily
+perform the conversion by using the :meth:`~.QNode.to_tf` method:
+
+>>> qnode1.to_tf()
 >>> qnode1
-<QNode: device='default.qubit', func=circuit, wires=2, interface=TensorFlow>
+<QNode: device='default.mixed', func=circuit1, wires=2, interface=TensorFlow>
+
+``qnode1()`` is now a TensorFlow-capable QNode, as well. Internally, the :meth:`~.QNode.to_tf`
+method uses the :func:`~.TFQNode` function to do the conversion.
```
Comment (albi3ro) [inline doc/introduction/interfaces/tf.rst:94]: Internally, the method uses [`~.interfaces.tf.TFInterface.apply`](https://github.com/PennyLaneAI/pennylane/blob/0965bb0c97c4de2dc2408751ca3a61dd47074264/pennylane/interfaces/tf.py#L225) to do the conversion.

```suggestion
method uses the :meth:`~.interfaces.tf.TFInterface.apply` function to do the conversion.
```

Check that actually works.

File: doc/introduction/interfaces/torch.rst [hunk 5]
```diff
@@ -21,66 +20,84 @@ new quantum node object.
...
+Let's say we change our mind and now want ``qnode1()`` to interface with PyTorch. We can easily
+perform the conversion by using the :meth:`~.QNode.to_torch` method:
+
+>>> qnode1.to_torch()
+>>> qnode1
+<QNode: device='default.mixed', func=circuit1, wires=2, interface=PyTorch>
+
+``qnode1()`` is now a PyTorch-capable QNode, as well. Internally, the
+:meth:`QNode.to_torch <qnode.QNode.to_torch>` method uses the
 :func:`TorchQNode <interfaces.torch.TorchQNode>` function to do the conversion.
```
Comment (albi3ro) [inline doc/introduction/interfaces/torch.rst:96]:
```suggestion
:meth:`TorchInterface.apply <~.interfaces.torch.TorchInterface.apply>` function to do the conversion.
```

Looks like this function needs to get updated here too.

**Review by albi3ro — CHANGES_REQUESTED** [review]
Looks great!

I left some comments about fixing the method names for the internal implementation, but after that it should be good to go :)

Above and beyond what I was initially wanting; a definite improvement.

File: doc/introduction/interfaces/tf.rst [hunk 6]
```diff
@@ -24,61 +23,78 @@ Tensorflow is imported as follows:
...
+Let's say we change our mind and now want ``qnode1()`` to interface with TensorFlow. We can easily
+perform the conversion by using the :meth:`~.QNode.to_tf` method:
+
+>>> qnode1.to_tf()
 >>> qnode1
-<QNode: device='default.qubit', func=circuit, wires=2, interface=TensorFlow>
+<QNode: device='default.mixed', func=circuit1, wires=2, interface=TensorFlow>
+
+``qnode1()`` is now a TensorFlow-capable QNode, as well. Internally, the :meth:`~.QNode.to_tf`
+method uses the :meth:`~.interfaces.tf.TFInterface.apply` function to do the conversion.
```
Comment (albi3ro) [inline doc/introduction/interfaces/tf.rst:94]:
```suggestion
method uses the :meth:`TFInterface.apply <~.interfaces.tf.TFInterface.apply>` function to do the conversion.
```

This was rendering as simply `apply`, so maybe we should try this instead. Neither was linking up to source code, but I'm not good enough at sphinx to figure out why.  Seems like maybe the interfaces are hidden internal objects that don't show up in the documentation.

I still don't think users need this information, especially if the method doesn't even show up in the documentation, but the text should at least be something more informative  than `apply`.

**Review by albi3ro — CHANGES_REQUESTED** [review]
Last tiny detail. 

Just make sure to check the build by going to the PR's checks, then `docs/readthedocs.org:pennylane` and clicking on `details`.

File: doc/introduction/interfaces/tf.rst (hunk 6 above)
Comment (josh146) [reply doc/introduction/interfaces/tf.rst:94]: Perhaps it could be moved to a call out/hidden dropdown for advanced users?

**Comment by josh146** [comment]
> @josh146 Just about! Should I add similar updates to the JAX interface page? I was holding off on that one until I got the PyTorch and TensorFlow ones right so I could just work off of those templates.

@ryanhill1 adding a similar update to the JAX interface page would be great (and this is something I know @albi3ro has also been thinking about!).

Best to keep it as a separate PR though. Usually it is best to have a separate PR per ‘task’/‘feature’/‘bug fix’, to help with the code review process.

**Comment by ryanhill1** [comment]
@josh146 @albi3ro Ok, I just deleted the lines about the internal use of the `apply` function in both the PyTorch and TensorFlow pages. I think you're right that it's not needed. I reviewed the builds in `docs/readthedocs.org:pennylane` and both files look good. And I also updated the CHANGELOG file. So everything should be ready for code review now!

File: .github/CHANGELOG.md [hunk 7]
```diff
@@ -344,14 +344,16 @@ random_mat2 = rng.standard_normal(3, requires_grad=False)
 
 * Removes occurrences of the deprecated device argument ``analytic`` from the documentation.
   [(#1261)](https://github.com/PennyLaneAI/pennylane/pull/1261)
+  
+* Updated PyTorch and TensorFlow interface introductions [(#1333)](https://github.com/PennyLaneAI/pennylane/pull/1333)
```
Comment (josh146) [inline .github/CHANGELOG.md:348]:
```suggestion
* Updated PyTorch and TensorFlow interface introductions.
  [(#1333)](https://github.com/PennyLaneAI/pennylane/pull/1333)
```
(just to match the convention of the other entries)

**Review by josh146 — COMMENTED** [review]
Awesome, thanks @ryanhill1! 💪 This looks great, and thanks for updating some of the (quite outdated) code snippets and links.

I did a quick skim through for rendering and ReST formatting, and all looks good on my end. I've tagged @albi3ro for a final review, and then should be ready for merging 🙂

**Review by albi3ro — APPROVED** [review]
Thanks! All good on my end :)

**Pull request merged**